liquidhaskell-boot/src/Language/Haskell/Liquid/Bare/Axiom.hs

```diff
@@ -346,6 +354,26 @@ makeAssumeType allowTC tce lmap dm sym mbT v def
     ref        = F.Reft (F.vv_, F.PAtom F.Eq (F.EVar F.vv_) le)
     mkErr s    = ErrHMeas (sourcePosSrcSpan $ loc sym) (pprint $ val sym) (PJ.text s)
     bbs        = filter isBoolBind xs
+
+    -- Before this modification reflected functions that were polymorphic were
+    -- getting compiled to monomorphic fixpint code.  As an example: 
+    -- id :: a -> a ... id x = x 
+    -- Was generating: 
+    -- define id (x : a#foobar) : a#foobar = { (x : a#foobar) }
+    -- Using FObj instead of a real type variable (FVar) This code solves the
+    -- issue by creating a sobritution that replaces those "fake" type variables
+    -- with actual ones.
+    -- define id (x : @-1) : a@-1 = { (x : a@-1) }
+    (tyVars, _) = Ghc.splitForAllTyCoVars τ
+    sortSub     = F.mkSortSubst $ zip (fmap F.symbol tyVars) (F.FVar <$> freeSort)
+    -- We need sorts that aren't polluted by rank-n types, we can't just look at
```

Comment:
I'm missing what the concern is with rank-n types. Though it is probably relevant and I need to become aware.

Reply:
suppose we have

```
def f (x : @0) : ... = { a x }
```

Where
```hs
a :: forall x. x -> (forall b . b -> ...)
```

During `a x` will be annotated with `a x : function(1, [@0; @0; ...]` Where the new forall (function) shadows the outsize type variable giving a wrong typing to the term that can lead to the best scenario crashes to infinite terms during unification (if was not fun at all to debug :)))